csrc/device_lower/analysis/tma.cpp

```diff
@@ -48,441 +50,735 @@ int64_t getCpAsyncBulkTensorSwizzleSize(TensorView* smem_tv) {
-    const ExprGroups& defs2 =
-        acyclicExprGroups(id_graph, id_graph.getDefinitions(box_g));
-    NVF_ERROR(
-        defs2.size() <= 1,
-        "Having multiple definitions of box group is not supported");
-    ExprGroup boxing_split = nullptr;
-    if (!defs2.empty() && defs2.front()->front()->isA<Split>()) {
-      boxing_split = defs2.front();
-    }
-    ValGroup partitioned_g =
-        (boxing_split != nullptr ? id_graph.inputGroups(boxing_split)[0]
-                                 : nullptr);
-    ValGroup tma_g = (partitioned_g != nullptr ? partitioned_g : box_g);
-
-    tma_groups.push_back(tma_g);
-    tma_g_to_box_g[tma_g] = box_g;
-    if (stride_g != nullptr) {
-      tma_g_to_tile_stride_g[tma_g] = {tile_g, stride_g};
-    }
-    if (partitioned_g != nullptr) {
-      partitioned_groups.insert(partitioned_g);
+ public:
+  AnalyzeBoxingSplit(
+      const std::unordered_set<ValGroup>& bulk_groups,
+      const std::unordered_set<ValGroup>& non_bulk_groups,
+      std::list<TMADim>& inferred_dims)
+      : bulk_groups_(bulk_groups),
+        non_bulk_groups_(non_bulk_groups),
+        inferred_dims_(inferred_dims) {}
+};
+
+std::tuple<
+    std::unordered_set<ValGroup>,
+    std::unordered_set<ValGroup>,
+    std::list<TMADim>>
+run(std::list<std::pair<ExprGroup, Direction>>& exprs,
+    TensorView* consumer_tv) {
+  ValGraph& id_graph = GpuLower::current()->tensorIndexer().traversalGraph();
+
+  std::unordered_set<ValGroup> bulk_groups;
+  std::unordered_set<ValGroup> nonbulk_groups;
+  for (auto id : consumer_tv->getLoopDomain()) {
+    if (id->getParallelType() == ParallelType::Bulk) {
+      bulk_groups.insert(id_graph.toGroup(id));
+    } else {
+      nonbulk_groups.insert(id_graph.toGroup(id));
     }
   }
 
-  // Stpe 3: Propagate from the gmen tensor's allocation domain to the TMA
-  // domain, compute the order, contiguity, and stride of partitioned
-  // ValGroups. Note that this order is meaningful, and it is the order that
-  // defines which is inner and which is outer. The strides are also meaningful,
-  // and they are the `globalStrides` of the `cuTensorMapEncodeTiled`. After
-  // propagation, `frontier` will be the TMA domain
+  std::list<TMADim> inferred_dims;
+
+  InferBulkGroups bulk_pass(bulk_groups);
+  InferNonBulkGroups nonbulk_pass(nonbulk_groups);
+  AnalyzeStridingSplit striding_split_pass(
+      bulk_groups, nonbulk_groups, inferred_dims);
+  AnalyzeBoxingSplit boxing_split_pass(
+      bulk_groups, nonbulk_groups, inferred_dims);
+
+  bool changed = true;
+  while (changed) {
+    changed = false;
+    changed = changed || bulk_pass.run(exprs);
+    changed = changed || nonbulk_pass.run(exprs);
+    changed = changed || striding_split_pass.run(exprs);
+    changed = changed || boxing_split_pass.run(exprs);
+  }
+  return {bulk_groups, nonbulk_groups, inferred_dims};
+}
 
+} // namespace infer_roles
+
+// View the allocation domain of the gmem tensor with given exprs, and compute
```

Comment:
Can you give a little more details here? For example, what does the viewing exactly mean? What exprs are typically given?

Reply:
This became clear when I finished reading the code in this file. It would have been clearer if there were some more comments upfront.

Reply:
rephrased this paragraph